neuro_flow/storage.py

```diff
@@ -729,3 +841,64 @@ def _attempt_from_json(
         number=init_data["number"],
         result=result,
     )
+
+
+def _mk_cache_uri(attempt: Attempt, task_id: FullID) -> URL:
+    return (
+        URL("storage:.flow")
+        / attempt.bake.project
+        / ".cache"
+        / attempt.bake.batch
+        / (".".join(task_id) + ".json")
+    )
+
+
+def _hash(val: Any) -> str:
+    hasher = hashlib.new("sha256")
+    data = json.dumps(val, sort_keys=True, default=_ctx_default)
+    hasher.update(data.encode("utf-8"))
+    return hasher.hexdigest()
+
+
+def _ctx_default(val: Any) -> Any:
+    if isinstance(val, BaseContext):
+        ret: Dict[str, Any] = {
+            "env": val.env,
+            "tags": val.tags,
+            "digest": val.digest,
+        }
+        if isinstance(val, ActionContext):
+            ret.update(
+                {
+                    "inputs": val.inputs,
+                }
+            )
+        elif isinstance(val, TaskContext):
+            ret.update(
+                {
+                    "needs": val.needs,
+                    "matrix": val.matrix,
+                    "strategy": val.strategy,
+                    "task": val.task,
+                }
+            )
+        return ret
+    elif dataclasses.is_dataclass(val):
+        return dataclasses.asdict(val)
+    elif isinstance(val, enum.Enum):
+        return val.value
+    elif isinstance(val, RemotePath):
+        return str(val)
+    elif isinstance(val, collections.abc.Set):
+        return sorted(val)
+    else:
+        raise TypeError(f"Cannot dump {val!r}")
+
+
+def _result_default(val: Any) -> Any:
```

Comment:
Can't find any usage of this function. Maybe I'm missing something?

Reply:
You are right.
It is an artifact of earlier implementation.